[LinearAlgebra] Cache `lbt_get_config()` result

staticfloat opened the pull request:
In the event that users want to call `lbt_get_config()` multiple times
(e.g. for runtime checks of which BLAS vendor is providing a symbol),
let's cache the value and clear it only when someone calls something
that would cause the config itself to change.

staticfloat committed:
[LinearAlgebra] Cache `lbt_get_config()` result

In the event that users want to call `lbt_get_config()` multiple times
(e.g. for runtime checks of which BLAS vendor is providing a symbol),
let's cache the value and clear it only when someone calls something
that would cause the config itself to change.

staticfloat commented:
A quick benchmark of this method on my Linux desktop shows that this speeds up `lbt_get_config()` from about 120 μs to 3ns.  It's not every day I get to speed a method up by a factor of 40,000!

tkf commented on stdlib/LinearAlgebra/src/lbt.jl:
```diff
@@ -171,9 +171,20 @@ function Base.show(io::IO, mime::MIME{Symbol("text/plain")}, lbt::LBTConfig)
+    return _cached_config[]
+end
+function _clear_cached_config()
+    global _cached_config[] = nothing
```
```suggestion
    _cached_config[] = nothing
```

tkf commented on stdlib/LinearAlgebra/src/lbt.jl:
```diff
@@ -171,9 +171,20 @@ function Base.show(io::IO, mime::MIME{Symbol("text/plain")}, lbt::LBTConfig)
 function lbt_get_config()
-    config_ptr = ccall((:lbt_get_config, libblastrampoline), Ptr{lbt_config_t}, ())
-    return LBTConfig(unsafe_load(config_ptr))
+    global _cached_config
```
```suggestion
```

tkf commented on stdlib/LinearAlgebra/src/lbt.jl:
```diff
@@ -171,9 +171,20 @@ function Base.show(io::IO, mime::MIME{Symbol("text/plain")}, lbt::LBTConfig)
+# In the event that users want to call `lbt_get_config()` multiple times (e.g. for
+# runtime checks of which BLAS vendor is providing a symbol), let's cache the value
+# and clear it only when someone calls something that would cause it to change.
+const _cached_config = Ref{Union{Nothing,LBTConfig}}(nothing)
```
If we want to make it "thead-safe", we might need the double-checked locking pattern or something here. That's tedious to do on a case-by-case basis so I guess we'd want it in Base.Threads.

tkf replied:
See #44387

tkf committed:
Use double-checked locking to cache LBT config (#44387)

* Use double-checked locking to cache LBT config

* Use Any field for lock-free load and store

* Revert "Use Any field for lock-free load and store"

This reverts commit c6acae5c9b460cc4c7807e612712b92adfaebb61.

The compiler already handled it:
https://github.com/JuliaLang/julia/pull/44387#discussion_r816730294

* (rerun)

* Add lock as a field of ConfigCache
